core/trino-main/src/main/java/io/trino/operator/window/pattern/LogicalIndexNavigation.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.window.pattern;
+
+import java.util.OptionalInt;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class LogicalIndexNavigation
+{
+    private final Set<Integer> labels;
+    private final boolean last;
+    private final boolean running;
+    private final int logicalOffset;
+    private final int physicalOffset;
+
+    public LogicalIndexNavigation(Set<Integer> labels, boolean last, boolean running, int logicalOffset, int physicalOffset)
+    {
+        this.labels = requireNonNull(labels, "labels is null");
+        this.last = last;
+        this.running = running;
+        checkArgument(logicalOffset >= 0, "logical offset must be >= 0, actual: " + logicalOffset);
+        this.logicalOffset = logicalOffset;
+        this.physicalOffset = physicalOffset;
+    }
+
+    /**
+     * This method is used for resolving positions during pattern matching. The `newLabel` is the label currently being matched.
+     * For the purpose of match evaluation, the new label is considered as matched in the next position after all `matchedLabels`.
+     */
+    public OptionalInt resolvePosition(int[] matchedLabels, int newLabel, int partitionStart, int partitionEnd, int patternStart)
+    {
+        OptionalInt relativePosition;
+        if (last) {
+            relativePosition = findLastAndBackwards(matchedLabels, newLabel);
+        }
+        else {
+            relativePosition = findFirstAndForward(matchedLabels, newLabel);
+        }
+        return adjustPosition(relativePosition, patternStart, partitionStart, partitionEnd);
+    }
+
+    // LAST(A.price, 3): find the last occurrence of label "A" and go 3 occurrences backwards
```

Comment:
Not sure I understand how the code in this method finds the last "A" given the loop condition:

```
if (labels.isEmpty() || labels.contains(matchedLabels[position])) {
```

What's `labels` in this context? Is that the list of defined labels? Or the label(s) to search for? If the latter, why are there multiple ones?

Reply:
I guess that the explanation in the comment is a bit misleading.
`labels` is a property of the navigation. It says, what kind of rows the navigation applies to (labels to search for):

- LAST(A.price, 3) => this is a navigation over rows with label `A`, so `labels = {A}`
- LAST(Union.price, 3) => this is a navigation over rows matching a union variable `Union`, so for `SUBSET Union = (A, B, C)`, we have `labels = {A, B, C}`
- LAST(price, 3) => this is a navigation over "universal pattern variable", which is effectively over all rows, no matter the assigned labels. In such case `labels = {}`